crates/noirc_evaluator/src/ssa/code_gen.rs

```diff
@@ -0,0 +1,1006 @@
+            }
+        }
+        None
+    }
+
+    pub fn dummy_id() -> arena::Index {
+        arena::Index::from_raw_parts(std::usize::MAX, 0)
+    }
+
+    pub fn dummy(&self) -> arena::Index {
+        IRGenerator::dummy_id()
+    }
+
+    pub fn get_object(&self, idx: arena::Index) -> Option<&node::NodeObj> {
+        self.nodes.get(idx)
+    }
+
+    pub fn get_mut_object(&mut self, idx: arena::Index) -> Option<&mut node::NodeObj> {
+        self.nodes.get_mut(idx)
+    }
+
+    fn get_object_type(&self, idx: arena::Index) -> node::ObjectType {
+        self.get_object(idx).unwrap().get_type()
+    }
+
+    //Returns the object value if it is a constant, None if not. TODO: handle types
+    pub fn get_as_constant(&self, idx: arena::Index) -> Option<FieldElement> {
+        if let Some(node::NodeObj::Const(c)) = self.get_object(idx) {
+            return Some(FieldElement::from_be_bytes_reduce(&c.value.to_bytes_be()));
+        }
+        None
+    }
+
+    pub fn get_mut_instruction(&mut self, idx: arena::Index) -> &mut node::Instruction {
+        if let Some(node::NodeObj::Instr(i)) = self.get_mut_object(idx) {
+            return i;
+        }
+        unreachable!("Index not found or not an instruction");
+    }
+
+    pub fn get_as_instruction(&self, idx: arena::Index) -> Option<&node::Instruction> {
+        if let Some(node::NodeObj::Instr(i)) = self.get_object(idx) {
+            return Some(i);
+        }
+        None
+    }
+
+    pub fn get_as_mut_instruction(&mut self, idx: arena::Index) -> Option<&mut node::Instruction> {
+        if let Some(node::NodeObj::Instr(i)) = self.get_mut_object(idx) {
+            return Some(i);
+        }
+        None
+    }
+
+    //todo handle errors
+    fn get_instruction(&self, idx: arena::Index) -> &node::Instruction {
+        if let Some(node::NodeObj::Instr(i)) = self.get_object(idx) {
+            return i;
+        }
+        unreachable!("Index not found or not an instruction");
+    }
+
+    pub fn get_variable(&self, idx: arena::Index) -> Result<&node::Variable, &str> {
+        //TODO proper error handling
+        match self.nodes.get(idx) {
+            Some(t) => match t {
+                node::NodeObj::Obj(o) => Ok(o),
+                _ => Err("Not an object"),
+            },
+            _ => Err("Invalid id"),
+        }
+    }
+
+    pub fn get_mut_variable(&mut self, idx: arena::Index) -> Result<&mut node::Variable, &str> {
+        //TODO proper error handling
+        match self.nodes.get_mut(idx) {
+            Some(t) => match t {
+                node::NodeObj::Obj(o) => Ok(o),
+                _ => Err("Not an object"),
+            },
+            _ => Err("Invalid id"),
+        }
+    }
+
+    pub fn try_into_instruction(&self, idx: arena::Index) -> Option<&node::Instruction> {
+        if let Some(node::NodeObj::Instr(i)) = self.get_object(idx) {
+            return Some(i);
+        }
+        None
+    }
+
+    pub fn try_into_mut_instruction(
```

Comment:
These 2 functions look like repeats of `get_as_instruction` above

Reply:
Done